src/Symfony/Component/Cache/Adapter/MemcacheAdapter.php

```diff
@@ -0,0 +1,144 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Cache\Adapter;
+
+use Symfony\Component\Cache\Exception\InvalidArgumentException;
+
+/**
+ * @author Rob Frawley 2nd <rmf@src.run>
+ */
+class MemcacheAdapter extends AbstractMemcacheAdapter
+{
+    /**
+     * Construct adapter by passing a \Memcache instance and an optional namespace and default cache entry ttl.
+     *
+     * @param \Memcache   $client
+     * @param string|null $namespace
+     * @param int         $defaultLifetime
+     */
+    public function __construct(\Memcache $client, $namespace = '', $defaultLifetime = 0)
+    {
+        parent::__construct($namespace, $defaultLifetime);
+        $this->client = $client;
+    }
+
+    /**
+     * Factory creation method that provides an instance of this adapter with a\Memcache client instantiated and setup.
+     *
+     * Valid DSN values include the following:
+     *  - memcached://localhost                  : Specifies only the host (defaults used for port and weight)
+     *  - memcached://example.com:1234           : Specifies host and port (defaults weight)
+     *  - memcached://example.com:1234?weight=50 : Specifies host, port, and weight (no defaults used)
+     *
+     * @param string|null $dsn
+     *
+     * @return MemcacheAdapter
+     */
+    public static function create($dsn = null)
+    {
+        if (!extension_loaded('memcache')) {
+            throw new InvalidArgumentException('Failed to create Memcache client due to missing "memcache" extension.');
+        }
+
+        $adapter = new static(new \Memcache());
+        $adapter->setup($dsn ? array($dsn) : array());
+
+        return $adapter;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function doSave(array $values, $lifetime)
+    {
+        $result = true;
+
+        foreach ($values as $id => $val) {
+            $result = $this->client->set($id, $val, null, $lifetime) && $result;
+        }
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function doFetch(array $ids)
+    {
+        foreach ($this->client->get($ids) as $id => $val) {
+            yield $id => $val;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function doHave($id)
+    {
+        return $this->client->get($id) !== false;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function doDelete(array $ids)
+    {
+        $remaining = array_filter($ids, function ($id) {
+            return false !== $this->client->get($id) && false === $this->client->delete($id);
+        });
+
+        return 0 === count($remaining);
+    }
+
+    protected function getAllIdsMatching($namespace)
```

Comment:
can be private when a trait is used instead of abstract (same in the other class)